### PR TITLE
Update: Replace several dependencies with lodash (fixes #5012)

### LINKED ISSUE #5012: Simplify dependency tree
https://github.com/eslint/eslint/issues/5000#issuecomment-172909102

> In my view, one of the flaws in ESLint setup is the use of "modular" lodash NPM packages.
> 
> You could get rid of these dependencies:
>
> ```
> "es6-map": "^0.1.3",
> "escape-string-regexp": "^1.0.2",
> "handlebars": "^4.0.5",
> "lodash.clonedeep": "^3.0.1",
> "lodash.merge": "^3.3.2",
> "lodash.omit": "^3.1.0",
> "object-assign": "^4.0.1",
> "xml-escape": "~1.0.0"
> by simply adding lodash@4 that has all of those features (and no sub-dependencies. The sub-dependency tree of these dependencies is huge).
> ``` 
>
> And there is no reason whatsoever to use "modular" lodash packages (given that ESLint is a development tool).> 

https://github.com/eslint/eslint/issues/5000#issuecomment-172990303

> @gajus if you're willing to go through and make those changes to use lodash@4 instead of the independent ones, we can take a look.

https://github.com/eslint/eslint/issues/5000#issuecomment-173221690

> I am on it.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/30047976-simplify-dependency-tree?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by gajus**

### CODE REVIEW

File: lib/formatters/html.js [hunk 1]
```diff
@@ -44,45 +48,79 @@ handlebars.registerHelper("renderText", function(totalErrors, totalWarnings) {
...
-});
+}
+
+/**
+ * @param {Array} messages Messages.
+ * @param {number} parentIndex Index of the parent HTML row.
+ * @returns {string} HTML string describing result messages.
+ */
+function renderMessages(messages, parentIndex) {
+    return _.map(messages, function(message) {
```
Comment (ilyavolodin) [inline lib/formatters/html.js]: I'd much rather use native map implementation.

File: package.json [hunk 2]
```diff
@@ -84,10 +77,10 @@
     "gh-got": "^2.2.0",
     "istanbul": "^0.4.0",
     "jsdoc": "^3.3.0-beta1",
-    "jsonlint": "^1.6.2",
     "leche": "^2.1.1",
     "linefix": "^0.1.1",
     "load-perf": "^0.2.0",
+    "lodash": "^4.0.0",
```
Comment (gyandeeps) [inline package.json]: Duplicate definition as its already present inside dependencies.

Comment (gajus) [inline package.json]: The reason it is duplicate is because both development and production environment is using lodash. Thats not a mistake.

Comment (gyandeeps) [inline package.json]: npm will warn you for this and it will always use dependency from `dependencies`.
`npm WARN package.json Dependency 'lodash' exists in both dependencies and devDependencies, using 'lodash@^4.0.0' from dependencies`

File: lib/formatters/html.js (hunk 1 above)
Comment (nzakas) [inline lib/formatters/html.js]: +1 We should use native whenever available.

**Comment by gyandeeps** [comment]
Can you plz squash your commits into 1 commit as per our guidelines: http://eslint.org/docs/developer-guide/contributing/pull-requests

**Comment by nzakas** [comment]
I forgot how much I dislike `_` as an identifier. Can we use `lodash` instead?

Also, I agree with @ilyavolodin, I think we should use the native equivalent whenever it's present. (`map`, `forEach`, etc.).

File: lib/formatters/html.js [hunk 3]
```diff
@@ -44,45 +48,79 @@ handlebars.registerHelper("renderText", function(totalErrors, totalWarnings) {
...
     // Iterate over results to get totals
     results.forEach(function(result) {
-        data.totalErrors += result.errorCount;
-        data.totalWarnings += result.warningCount;
+        totalErrors += result.errorCount;
+        totalWarnings += result.warningCount;
     });
 
-    return handlebars.compile(template)(data);
+    // console.log('results', results);
```
Comment (gyandeeps) [inline lib/formatters/html.js]: Can we remove this?

**Comment by gajus** [comment]
> [@nzakas] Also, I agree with @ilyavolodin, I think we should use the native equivalent whenever it's present. (map, forEach, etc.).

Before we proceed further, please review https://github.com/eslint/eslint/issues/5017.

File: lib/formatters/html.js [hunk 4]
```diff
@@ -44,45 +48,77 @@ handlebars.registerHelper("renderText", function(totalErrors, totalWarnings) {
...
     return 0;
-});
+}
 
 /**
- * Get the HTML row content based on the severity of the message
- * @param {int} severity Severity of the message
- * @returns {string} The generated HTML row
+ * @param {Array} messages Messages.
+ * @param {number} parentIndex Index of the parent HTML row.
```
Comment (nzakas) [inline lib/formatters/html.js]: `{int}`. We differentiate between ints and floats.

**Comment by zandaqo** [comment]
`Object.assign` is supported in Node 4+ and, If I'm not mistaken, there was a talk here about dropping support for older node versions in few months. In the light of it, is it really necessary to replace all assigns with the _.assign right now? 

Also, if you go with this, wouldn't it be clearer to just re-assign `assign = require('object-assing') to `_.assing` instead of changing each use to `_.assign`? It would also make it easier to switch to the native implementation later.

File: Makefile.js [hunk 5]
```diff
@@ -563,10 +578,7 @@ target.lint = function() {
     }
 
     echo("Validating JSON Files");
-    lastReturn = nodeCLI.exec("jsonlint", "-q -c", JSON_FILES);
-    if (lastReturn.code !== 0) {
-        errors++;
-    }
+    _.forEach(JSON_FILES, validateJsonFile);
```
Comment (mysticatea) [inline Makefile.js]: I like the tolerant behavior of original.

Comment (nzakas) [inline Makefile.js]: Agreed.

Comment (gajus) [inline Makefile.js]: The original was crashing the makefile.js.

File: lib/rules/no-unmodified-loop-condition.js [hunk 6]
```diff
@@ -261,10 +261,10 @@ module.exports = function(context) {
 
             for (var j = 0; j < condition.groups.length; ++j) {
                 var be = condition.groups[j];
-                var group = groupMap.get(be);
+                var group = groupMap[be];
```
Comment (mysticatea) [inline lib/rules/no-unmodified-loop-condition.js]: `be` is not strings.

Comment (gajus) [inline lib/rules/no-unmodified-loop-condition.js]: Then it is a test problem to begin with. Can you update the tests to make this fail?

Comment (nzakas) [inline lib/rules/no-unmodified-loop-condition.js]: ES6 maps can have nonstring keys.

Comment (gajus) [inline lib/rules/no-unmodified-loop-condition.js]: Thats fine. I am only saying that if this is a relevant use case, then there should be a test that would cover it. I will re-implement it using `_.find` and an array of collections.

Comment (mysticatea) [inline lib/rules/no-unmodified-loop-condition.js]: Indeed the tests are lacking, though, I don't think we need replace the `Map` object by `lodash`. Because the `Map` object is a native object of ECMAScript. It's supported on Node since v0.12. We can just remove `es6-map` when we drop v0.10 support. And `escope` is using `es6-map`, so we cannot remove `es6-map` from dependencies even if removed from here.

File: package.json [hunk 7]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "eslint",
-  "version": "2.0.0-rc.0",
+  "version": "2.0.0-beta.1",
```
Comment (gyandeeps) [inline package.json]: I think you need to rebase your branch with origin/master as this change is not correct.

**Comment by gyandeeps** [comment]
Please rebase as
* recently more code was added to master which used lodash, i dont see changes to that code (auto config code)
* Currently you are overwriting some of the changes in trunk.

File: lib/formatters/html.js [hunk 8]
```diff
@@ -44,45 +48,77 @@ handlebars.registerHelper("renderText", function(totalErrors, totalWarnings) {
...
     }
     return 0;
-});
+}
 
 /**
- * Get the HTML row content based on the severity of the message
- * @param {int} severity Severity of the message
- * @returns {string} The generated HTML row
+ * @param {Array} messages Messages.
```
Comment (gyandeeps) [inline lib/formatters/html.js]: Why was the description of this function removed?

File: lib/formatters/html.js [hunk 9]
```diff
@@ -44,45 +48,77 @@ handlebars.registerHelper("renderText", function(totalErrors, totalWarnings) {
...
+            columnNumber: columnNumber,
+            severityNumber: message.severity,
+            severityName: message.severity === 1 ? "Warning" : "Error",
+            message: message.message,
+            ruleId: message.ruleId
+        });
+    }).join("\n");
+}
+
+/**
```
Comment (gyandeeps) [inline lib/formatters/html.js]: function description missing as we try to put descriptions in jsdoc for each function.
Yes, our static analysis should check this but we are working on it to activate more rules for all the eslint repos.

File: package.json [hunk 10]
```diff
@@ -100,9 +92,10 @@
     "phantomjs": "1.9.18",
     "phantomjs-polyfill": "0.0.1",
     "proxyquire": "^1.0.0",
+    "rewire": "^2.3.4",
```
Comment (gyandeeps) [inline package.json]: Why is this needed?

File: package.json [hunk 11]
```diff
@@ -100,9 +92,10 @@
...
     "semver": "^5.0.3",
     "shelljs-nodecli": "~0.1.0",
-    "sinon": "^1.17.2",
+    "sinon": "1.17.2",
```
Comment (gyandeeps) [inline package.json]: Why is this change necessary?

File: package.json (hunk 7 above)
Comment (gajus) [inline package.json]: Something is off with the latest merge.

Comment (gajus) [inline package.json]:
```
eslint git:(issue5012) git fetch upstream
eslint git:(issue5012) git rebase upstream/master
Current branch issue5012 is up to date.
```

Will need to correct these mistakes by hand it appears.

File: package.json (hunk 11 above)
Comment (gajus) [inline package.json]: Thank you for reviewing and sorry for the mess. Must have rushed something with the last merge.

File: package.json [hunk 12]
```diff
@@ -40,28 +40,22 @@
...
-    "escope": "^3.4.0",
-    "espree": "^3.0.0",
+    "escope": "^3.3.0",
+    "espree": "3.0.0",
     "estraverse": "^4.1.1",
     "estraverse-fb": "^1.3.1",
     "esutils": "^2.0.2",
     "file-entry-cache": "^1.1.1",
-    "glob": "^6.0.4",
+    "glob": "^5.0.14",
```
Comment (gyandeeps) [inline package.json]: woops.., another one ...

File: package.json [hunk 13]
```diff
@@ -40,28 +40,22 @@
     "debug": "^2.1.1",
     "doctrine": "^1.1.0",
     "es6-map": "^0.1.3",
-    "escape-string-regexp": "^1.0.2",
-    "escope": "^3.4.0",
-    "espree": "^3.0.0",
+    "escope": "^3.3.0",
+    "espree": "3.0.0",
```
Comment (gyandeeps) [inline package.json]: here also, the version number are not correct for espree and escope vs the master.

File: lib/formatters/html.js [hunk 14]
```diff
@@ -44,45 +48,83 @@ handlebars.registerHelper("renderText", function(totalErrors, totalWarnings) {
...
-});
+}
 
 /**
- * Get the HTML row content based on the severity of the message
- * @param {int} severity Severity of the message
- * @returns {string} The generated HTML row
+ * Get HTML (table rows) describing the messages.
+ * @param {Array} messages Messages.
+ * @param {number} parentIndex Index of the parent HTML row.
```
Comment (gyandeeps) [inline lib/formatters/html.js]: as @nzakas mentioned, we use `int` for cases like this?

Comment (gajus) [inline lib/formatters/html.js]: He did. Updated all.

**Comment by gajus** [comment]
Are there still any changes that you await from me?

File: Makefile.js [hunk 15]
```diff
@@ -60,7 +61,7 @@ var NODE = "node ", // intentional extra space
     MAKEFILE = "./Makefile.js",
     PACKAGE = "./package.json",
     JS_FILES = find("lib/").filter(fileType("js")).join(" "),
-    JSON_FILES = find("conf/").filter(fileType("json")).join(" ") + " .eslintrc",
+    JSON_FILES = find("conf/").filter(fileType("json")),
```
Comment (ilyavolodin) [inline Makefile.js]: Looks like this isn't going to validate `.eslintrc` file anymore, which is also json.

File: tests/lib/formatters/jslint-xml.js [hunk 16]
```diff
@@ -128,7 +128,7 @@ describe("formatter:jslint-xml", function() {
 
         it("should return a string in JSLint XML format with 1 issue in 1 file", function() {
             var result = formatter(code);
-            assert.equal(result, "<?xml version=\"1.0\" encoding=\"utf-8\"?><jslint><file name=\"foo.js\"><issue line=\"5\" char=\"10\" evidence=\"foo\" reason=\"Unexpected &lt;&amp;&quot;&apos;&gt; foo. (foo)\" /></file></jslint>");
+            assert.equal(result, "<?xml version=\"1.0\" encoding=\"utf-8\"?><jslint><file name=\"foo.js\"><issue line=\"5\" char=\"10\" evidence=\"foo\" reason=\"Unexpected &lt;&amp;&quot;&#39;&gt; foo. (foo)\" /></file></jslint>");
```
Comment (ilyavolodin) [inline tests/lib/formatters/jslint-xml.js]: This seems like an unrelated change. I'm not sure if this is right

**Comment by ilyavolodin** [comment]
@gajus Sorry, slipped through the cracks. It looks like one file is missing from json linting. Other then two comments above, everything else looks good. Thanks!

File: tests/lib/formatters/jslint-xml.js (hunk 16 above)
Comment (gajus) [inline tests/lib/formatters/jslint-xml.js]: In contrast to [`xml-escape`](https://www.npmjs.com/package/xml-escape), [`_.escape`](https://lodash.com/docs#escape) uses numeric code character entities instead of named ones. Thus the change.

File: Makefile.js (hunk 15 above)
Comment (gajus) [inline Makefile.js]: @ilyavolodin `.eslintrc` file does not exist. Your earlier JSON validation script was flawed. I made a comment about this in one of the earlier commit amendments.

**Comment by ilyavolodin** [comment]
In that case. LGTM

**Comment by mourner** [comment]
> Also, I agree with @ilyavolodin, I think we should use the native equivalent whenever it's present. (map, forEach, etc.).

@ilyavolodin @nzakas would you accept a PR that gets rid of lodash altogether? I feel like using the whole package is really an overkill, given that all of the usage can be replaced with just a few lines of simple native code.

**Comment by gajus** [comment]
> @ilyavolodin @nzakas would you accept a PR that gets rid of lodash altogether? I feel like using the whole package is really an overkill, given that all of the usage can be replaced with just a few lines of simple native code.

Can you state a reason for wanting to get rid of lodash in the first place?

**Comment by gajus** [comment]
Note, that lodash is already a dependency of ESLint (before this commit). In fact, multiple versions:

```
└─┬ eslint@1.10.3
  ├─┬ chalk@1.1.1
  │ ├── ansi-styles@2.1.0
  │ ├── has-ansi@2.0.0
  │ ├── strip-ansi@3.0.0
  │ └── supports-color@2.0.0
  ├─┬ concat-stream@1.5.1
  │ ├── inherits@2.0.1
  │ ├─┬ readable-stream@2.0.5
  │ │ ├── core-util-is@1.0.2
  │ │ ├── process-nextick-args@1.0.6
  │ │ ├── string_decoder@0.10.31
  │ │ └── util-deprecate@1.0.2
  │ └── typedarray@0.0.6
  ├─┬ debug@2.2.0
  │ └── ms@0.7.1
  ├─┬ doctrine@0.7.2
  │ ├── esutils@1.1.6
  │ └── isarray@0.0.1
  ├── escape-string-regexp@1.0.4
  ├─┬ escope@3.4.0
  │ ├─┬ es6-map@0.1.3
  │ │ ├── d@0.1.1
  │ │ ├── es5-ext@0.10.11
  │ │ ├── es6-iterator@2.0.0
  │ │ ├── es6-set@0.1.4
  │ │ ├── es6-symbol@3.0.2
  │ │ └── event-emitter@0.3.4
  │ ├── es6-weak-map@2.0.1
  │ └─┬ esrecurse@3.1.1
  │   └── estraverse@3.1.0
  ├── espree@2.2.5
  ├── estraverse@4.1.1
  ├── estraverse-fb@1.3.1
  ├── esutils@2.0.2
  ├─┬ file-entry-cache@1.2.4
  │ └─┬ flat-cache@1.0.10
  │   ├─┬ del@2.2.0
  │   │ ├─┬ globby@4.0.0
  │   │ │ ├─┬ array-union@1.0.1
  │   │ │ │ └── array-uniq@1.0.2
  │   │ │ ├── arrify@1.0.1
  │   │ │ └── glob@6.0.4
  │   │ ├── is-path-cwd@1.0.0
  │   │ ├─┬ is-path-in-cwd@1.0.0
  │   │ │ └── is-path-inside@1.0.0
  │   │ ├── pify@2.3.0
  │   │ ├─┬ pinkie-promise@2.0.0
  │   │ │ └── pinkie@2.0.4
  │   │ └─┬ rimraf@2.5.1
  │   │   └── glob@6.0.4
  │   ├── graceful-fs@4.1.3
  │   ├── read-json-sync@1.1.1
  │   └── write@0.2.1
  ├─┬ glob@5.0.15
  │ ├─┬ inflight@1.0.4
  │ │ └── wrappy@1.0.1
  │ └── once@1.3.3
  ├── globals@8.18.0
  ├─┬ handlebars@4.0.5
  │ ├── async@1.5.2
  │ ├── optimist@0.6.1
  │ ├─┬ source-map@0.4.4
  │ │ └── amdefine@1.0.0
  │ └─┬ uglify-js@2.6.1
  │   ├── async@0.2.10
  │   ├── source-map@0.5.3
  │   ├── uglify-to-browserify@1.0.2
  │   └─┬ yargs@3.10.0
  │     ├── camelcase@1.2.1
  │     ├─┬ cliui@2.1.0
  │     │ ├─┬ center-align@0.1.3
  │     │ │ ├─┬ align-text@0.1.4
  │     │ │ │ ├─┬ kind-of@3.0.2
  │     │ │ │ │ └── is-buffer@1.1.2
  │     │ │ │ ├── longest@1.0.1
  │     │ │ │ └── repeat-string@1.5.2
  │     │ │ └── lazy-cache@1.0.3
  │     │ ├── right-align@0.1.3
  │     │ └── wordwrap@0.0.2
  │     ├── decamelize@1.1.2
  │     └── window-size@0.1.0
  ├─┬ inquirer@0.11.4
  │ ├── ansi-escapes@1.1.1
  │ ├── ansi-regex@2.0.0
  │ ├─┬ cli-cursor@1.0.2
  │ │ └─┬ restore-cursor@1.0.1
  │ │   ├── exit-hook@1.1.1
  │ │   └── onetime@1.1.0
  │ ├── cli-width@1.1.1
  │ ├── figures@1.4.0
  │ ├── lodash@3.10.1
  │ ├─┬ readline2@1.0.1
  │ │ ├─┬ code-point-at@1.0.0
  │ │ │ └── number-is-nan@1.0.0
  │ │ ├── is-fullwidth-code-point@1.0.0
  │ │ └── mute-stream@0.0.5
  │ ├── run-async@0.1.0
  │ ├── rx-lite@3.1.2
  │ ├── string-width@1.0.1
  │ └── through@2.3.8
  ├─┬ is-my-json-valid@2.12.4
  │ ├── generate-function@2.0.0
  │ ├─┬ generate-object-property@1.2.0
  │ │ └── is-property@1.0.2
  │ ├── jsonpointer@2.0.0
  │ └── xtend@4.0.1
  ├─┬ is-resolvable@1.0.0
  │ └── tryit@1.0.2
  ├─┬ js-yaml@3.4.5
  │ ├─┬ argparse@1.0.6
  │ │ └── sprintf-js@1.0.3
  │ └── esprima@2.7.2
  ├─┬ json-stable-stringify@1.0.1
  │ └── jsonify@0.0.0
  ├─┬ lodash.clonedeep@3.0.2
  │ ├─┬ lodash._baseclone@3.3.0
  │ │ ├── lodash._baseassign@3.2.0
  │ │ └── lodash._basefor@3.0.3
  │ └── lodash._bindcallback@3.0.1
  ├─┬ lodash.merge@3.3.2
  │ ├── lodash._arraycopy@3.0.0
  │ ├── lodash._arrayeach@3.0.0
  │ ├─┬ lodash._createassigner@3.1.1
  │ │ └── lodash._isiterateecall@3.0.9
  │ ├── lodash._getnative@3.9.1
  │ ├── lodash.isarguments@3.0.6
  │ ├── lodash.isarray@3.0.4
  │ ├── lodash.isplainobject@3.2.0
  │ ├── lodash.istypedarray@3.0.4
  │ ├── lodash.keys@3.1.2
  │ ├── lodash.keysin@3.0.8
  │ └─┬ lodash.toplainobject@3.0.0
  │   └── lodash._basecopy@3.0.1
  ├─┬ lodash.omit@3.1.0
  │ ├── lodash._arraymap@3.0.0
  │ ├─┬ lodash._basedifference@3.0.3
  │ │ ├── lodash._baseindexof@3.1.0
  │ │ ├── lodash._cacheindexof@3.0.2
  │ │ └── lodash._createcache@3.1.2
  │ ├── lodash._baseflatten@3.1.4
  │ ├── lodash._pickbyarray@3.0.2
  │ ├── lodash._pickbycallback@3.0.0
  │ └── lodash.restparam@3.6.1
  ├─┬ minimatch@3.0.0
  │ └─┬ brace-expansion@1.1.2
  │   ├── balanced-match@0.3.0
  │   └── concat-map@0.0.1
  ├─┬ mkdirp@0.5.1
  │ └── minimist@0.0.8
  ├── object-assign@4.0.1
  ├─┬ optionator@0.6.0
  │ ├── deep-is@0.1.3
  │ ├── fast-levenshtein@1.0.7
  │ ├── levn@0.2.5
  │ ├── prelude-ls@1.1.2
  │ ├── type-check@0.3.2
  │ └── wordwrap@0.0.3
  ├── path-is-absolute@1.0.0
  ├── path-is-inside@1.0.1
  ├── shelljs@0.5.3
  ├── strip-json-comments@1.0.4
  ├── text-table@0.2.0
  ├─┬ user-home@2.0.0
  │ └── os-homedir@1.0.1
  └── xml-escape@1.0.0
```

**Comment by gajus** [comment]
This commit removes all of this mess:

```
  ├─┬ lodash.clonedeep@3.0.2
  │ ├─┬ lodash._baseclone@3.3.0
  │ │ ├── lodash._baseassign@3.2.0
  │ │ └── lodash._basefor@3.0.3
  │ └── lodash._bindcallback@3.0.1
  ├─┬ lodash.merge@3.3.2
  │ ├── lodash._arraycopy@3.0.0
  │ ├── lodash._arrayeach@3.0.0
  │ ├─┬ lodash._createassigner@3.1.1
  │ │ └── lodash._isiterateecall@3.0.9
  │ ├── lodash._getnative@3.9.1
  │ ├── lodash.isarguments@3.0.6
  │ ├── lodash.isarray@3.0.4
  │ ├── lodash.isplainobject@3.2.0
  │ ├── lodash.istypedarray@3.0.4
  │ ├── lodash.keys@3.1.2
  │ ├── lodash.keysin@3.0.8
  │ └─┬ lodash.toplainobject@3.0.0
  │   └── lodash._basecopy@3.0.1
  ├─┬ lodash.omit@3.1.0
  │ ├── lodash._arraymap@3.0.0
  │ ├─┬ lodash._basedifference@3.0.3
  │ │ ├── lodash._baseindexof@3.1.0
  │ │ ├── lodash._cacheindexof@3.0.2
  │ │ └── lodash._createcache@3.1.2
  │ ├── lodash._baseflatten@3.1.4
  │ ├── lodash._pickbyarray@3.0.2
  │ ├── lodash._pickbycallback@3.0.0
  │ └── lodash.restparam@3.6.1
```

I will write a PR to bump `inquirer` `lodash` dependency. With that in place, ESLint download time ought to be faster than ever.

**Comment by gajus** [comment]
Turns out that the issue is with ESLint (https://github.com/SBoudrias/Inquirer.js/issues/333). Inquirer.js is already using an up to date lodash version. This requires bumping ESLint `inquirer` dependency version.

**Comment by mourner** [comment]
Without Lodash, NPM dependency trees throughout the whole JS ecosystem would be many times less bloated.

Lodash is unnecessary in Inquirer as well — a bunch of `_.extend` and redundant calls like `_.isString`, etc. Lodash was also present in `argparse` without any reason, but it got a [cleansing PR](https://github.com/nodeca/argparse/pull/91) already. :)

**Comment by gajus** [comment]
If size is the only argument, then it is irrelevant. Bundling dependencies (https://github.com/eslint/eslint/issues/5013) is one way to stop worrying about the size of the dependencies.

**Comment by mourner** [comment]
@gajus bundling dependencies and getting rid of lodash are not mutually exclusive tasks — doing both will lead to extraordinarily fast install times. I'm installing eslint very often (I write a lot of JS stuff) so any gain here would benefit me. In addition, getting rid of bloat is something I like to practice — it pays off in the long run.

**Comment by gajus** [comment]
My priority is to have the code base (the surface, the actual code that I need to maintain) as little as possible. Using standard utility functions (such as those that lodash provide) help to achieve this.

Size of the end package is off no importance as long as it does not impact the install time. Bundled dependencies solve the latter issue. In the days of fast internet and an average disk size of 590GB (http://www.eteknix.com/average-hard-drive-size-is-now-590gb/), deciding on whether to include 4MB (~250 gziped) library or not makes no sense.

**Comment by gajus** [comment]
Added a benchmark to support my case, https://github.com/eslint/eslint/issues/5013#issuecomment-181636428.
